src/Components/Web.JS/src/Platform/Circuits/CircuitManager.ts

```diff
@@ -233,43 +285,139 @@ export class CircuitManager implements DotNet.DotNetCallDispatcher {
+
+      // Indicate to the server that we want to pause the circuit.
+      // The server will initiate the pause on the circuit associated with the current connection.
+      const paused = await this._connection!.invoke<boolean>('PauseCircuit')!;
+      this._pausingState.complete(paused);
+    } catch (error) {
+      this._logger.log(LogLevel.Error, `Failed to pause the circuit: ${error}`);
+      this._pausingState.fail(error);
+    }
+
+    await this.disconnect();
+
+    return pausingPromise;
+  }
+
   public async resume(): Promise<boolean> {
     if (!this._circuitId) {
-      throw new Error('Method not implemented.');
+      this._logger.log(LogLevel.Error, 'Circuit host not initialized.');
+      throw new Error('Circuit host not initialized.');
     }
 
-    // When we get here we know the circuit is gone for good.
-    // Signal that we are about to start a new circuit so that
-    // any existing handlers can perform the necessary cleanup.
-    for (const handler of this._options.circuitHandlers) {
-      if (handler.onCircuitClosed) {
-        handler.onCircuitClosed();
-      }
+    if (this._disconnectingState.isInprogress()) {
+      // The handler might run before the `onclose` handler gets a chance to run.
+      this._logger.log(LogLevel.Trace, 'Circuit is disconnecting, cannot resume.');
+      await this._disconnectingState.currentProgress();
     }
 
-    const newCircuitId = await this._connection!.invoke<string>(
-      'ResumeCircuit',
-      this._circuitId,
-      navigationManagerFunctions.getBaseURI(),
-      navigationManagerFunctions.getLocationHref(),
-      '[]',
-      ''
-    );
-    if (!newCircuitId) {
+    if (this._pausingState.isInprogress()) {
+      this._logger.log(LogLevel.Trace, 'Waiting for the circuit to finish pausing...');
       return false;
     }
 
-    this._circuitId = newCircuitId;
-    this._renderQueue = new RenderQueue(this._logger);
-    for (const handler of this._options.circuitHandlers) {
-      if (handler.onCircuitOpened) {
-        handler.onCircuitOpened();
-      }
+    if (!this._pausingState.lastValue()) {
+      // If the circuit is not paused, we cannot resume it.
+      this._logger.log(LogLevel.Trace, 'Circuit is not paused.');
+      return false;
     }
 
-    this._options.reconnectionHandler!.onConnectionUp();
-    this._componentManager.onComponentReload?.();
-    return true;
+    if (this._connection!.state !== HubConnectionState.Connected) {
+      this._logger.log(LogLevel.Trace, 'Reestablishing SignalR connection...');
+      this._connection = await this.startConnection();
+    }
+
+    if (this._resumingState.isInprogress()) {
+      // If we are already resuming, wait for the current resume operation to complete.
+      this._logger.log(LogLevel.Trace, 'Waiting for the circuit to finish resuming...');
+      return this._resumingState.currentProgress();
```

Comment:
could resuming fail at this point ? We return from the method and reconnect code below would not run.